core/src/main/java/org/elasticsearch/index/engine/EngineConfig.java

```diff
@@ -303,4 +342,7 @@ public OpenMode getOpenMode() {
         OPEN_INDEX_AND_TRANSLOG;
     }
 
+    public interface EngineCreationListener {
```

Comment:
I don't think we should do this. Nothing should get access to an instance before the ctor returns. I will comment inside IndexShard how to pass around things...